src/printer.js

```diff
@@ -3837,19 +3828,38 @@ function printJSXElement(path, options, print) {
     return openingLines;
   }
 
+  // Convert `{" "}` to text nodes containing a space.
+  // This makes it easy to turn them into `jsxWhitespace` which
+  // can then print as either a space or `{" "}` when breaking.
+  n.children = n.children.map(child => {
```

Comment:
Usually, the strategy is instead of mutating the elements upfront, to update all the call sites of children to return you a list with your modifications. I'm not sure how many uses of children there are though, it may be tricky.